relayer/middleware/missedVaas/missedVaasV3.ts

```diff
@@ -0,0 +1,391 @@
+        emitterAddress: filter.emitterFilter.emitterAddress,
+      };
+    });
+    startMissedVaasWorkers(filters, redisPool, app.processVaa, opts);
+    registerEventListeners(app, redisPool, opts);
+  }, 100);
+}
+
+async function startMissedVaasWorkers(
+  filters: FilterIndentifier[],
+  redisPool: Pool<Cluster | Redis>,
+  processVaa: ProcessVaaFn,
+  opts: MissedVaaOpts,
+) {
+  const metrics: any = opts.registry ? initMetrics(opts.registry) : {};
+
+  if (opts.storagePrefix) {
+    const startTime = Date.now();
+    const scannedKeys = await updateSeenSequences(filters, redisPool, opts);
+    const elapsedTime = Date.now() - startTime;
+    opts.logger?.info(`Scanned ${scannedKeys} keys in ${elapsedTime}ms`);
+    metrics.workersWarmupDuration?.observe(elapsedTime);
+  }
+
+  while (true) {
+    opts.logger.info(`Missed VAA middleware run starting...`);
+    await mapConcurrent(
+      filters,
+      async filter => {
+        const { emitterChain, emitterAddress } = filter;
+        let vaasFound = 0;
+
+        opts.logger?.debug(
+          `Checking for missed vaas for emitterChain/emitterAddress: ${emitterChain}/${emitterAddress}`
+        );
+        const startTime = Date.now();
+        try {
+          vaasFound = await checkForMissedVaas(filter, redisPool, processVaa, opts);
+
+          metrics.workerSuccessfulRuns?.labels({ emitterChain, emitterAddress }).inc();
+          if (vaasFound > 0) metrics.recoveredVaas?.labels({ emitterChain, emitterAddress }).inc(vaasFound);
+        } catch (error) {
+          metrics.workerFailedRuns?.labels({ emitterChain, emitterAddress }).inc();
+
+          opts.logger?.error(
+            `Error checking for missed vaas for filter: ${JSON.stringify(filter)}. Error: ${error.toString()}`,
+          );
+        } finally {
+          const endTime = Date.now();
+          metrics.workersRunDuration?.labels({ emitterChain, emitterAddress }).observe(endTime - startTime);
+        }
+        opts.logger?.debug(
+          `Finished missed vaas check for emitterChain/emitterAddress: ${emitterChain}/${emitterAddress}. Found: ${vaasFound}`
+        );
+      },
+      opts.missedVaasConcurrency || 1,
+    );
+
+    // TODO: if possible handle this in a more event driven way (intervals + locks on a per chain basis)
+    sleep(opts.checkForMissedVaasEveryMs || 30_000);
+  }
+}
+
+async function updateSeenSequences(
+  filters: FilterIndentifier[],
+  redisPool: Pool<Cluster | Redis>,
+  opts: MissedVaaOpts,
+) {
+  let scannedKeys = 0;
+  for (const filter of filters) {
+    const { emitterChain, emitterAddress } = filter;
+    opts.logger?.info(`Updating seen sequences for filter: ${emitterChain}/${emitterAddress}`);
+    const redis = await redisPool.acquire();
+    scannedKeys += await scanNextBatchAndUpdateSeenSequences(redis, filter, opts.storagePrefix);
+  }
+  return scannedKeys;
+}
+
+const CURSOR_IDENTIFIER = "0";
+async function scanNextBatchAndUpdateSeenSequences(
```

Comment:
I just think more naturally about recursive solutions often, specially when iterating streams or other constructs where you don't necessarily know the length of the thing you are iterating. I'm ok with changing the approach if you prefer. You think it's harder to read?

Reply:
I generally try to avoid as much as possible changing the values of variables. IMO the more variables you change the value of, the harder the program gets to read, since the programmer has to keep track of more stuff in it's head.